### PR TITLE
Hide free trial banner when there's no internet connection

### LINKED ISSUE #8592: [Free Trial] [Test Plan Request] Adjust Free Trial banner behavior when offline
Summary
==========
If a merchant is viewing a Free Trial site without internet connectivity, and they tap on the Upgrade Now link/button, on returning to the app, the banner no longer appears even though they weren't able to upgrade.

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #8592

### Description

This PR adds behaviour of hiding the Free Trial banner when user has no internet connection.
### Testing instructions

With a site in Free Trial, turn on and off internet connection. Assert that the Free Trial banner doesn't appear when user has no internet connection.

### Images/gif

https://user-images.githubusercontent.com/5845095/230932771-c67a4bce-cad4-4577-a56d-97ffcd066f0f.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/plans/trial/DetermineTrialStatusBarStateTest.kt [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.stub
+import org.wordpress.android.fluxc.model.SiteModel
+import java.time.Period
+import java.time.ZoneId
+import java.time.ZonedDateTime
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class DetermineTrialStatusBarStateTest : BaseUnitTest() {
```
Comment (atorresveiga) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/plans/trial/DetermineTrialStatusBarStateTest.kt:32]: Nice!

**Comment by atorresveiga** [comment]
Great work, @wzieba!  The code looks good to me. Thanks for documenting all the different scenarios with the unit tests 🚀

I've found an issue. If I turn off/on the internet on my device, the trial banner disappears. Here is a video:

https://user-images.githubusercontent.com/18119390/231464620-823bb83e-fabb-48c7-8bf3-6d073798fef6.mp4

Is this expected?

**Comment by wzieba** [comment]
Thanks @atorresveiga !

> Is this expected?

No, this is not expected, thank you for the report! I haven't seen such behavior on my device :/. I'll try to reproduce it though somehow.

**Comment by atorresveiga** [comment]
@wzieba From your video, I see that you change the internet connection with the app in foreground. Could it be that in my case, the issue is that the app goes to background while changing the connection 🤔 
> I haven't seen such behavior on my device

I used a Pixel 6, in case it helps

**Comment by wzieba** [comment]
I can't reproduce this issue @atorresveiga :( how long have you been waiting for the banner to show up? There's a slight delay because we re-fetch data about the Free Trial plan from the remote - maybe that was the problem? Can I ask you to reproduce this issue one more time but wait for 10 seconds after enabling the internet connection?

**Comment by wzieba** [comment]
hi @atorresveiga 👋 just a friendly reminder about this issue, as it'd be wonderful to merge it before Friday 🙂.

**Comment by atorresveiga** [comment]
>how long have you been waiting for the banner to show up? There's a slight delay because we re-fetch data about the Free Trial plan from the remote - maybe that was the problem? Can I ask you to reproduce this issue one more time but wait for 10 seconds after enabling the internet connection?

Hi @wzieba!
I re-tested the scenario waiting for 10 seconds, and the same issue still happens to me. I opened Flipper and noted that after I enabled the internet again, the app was not calling to `public-api.wordpress.com/rest/v1.3/sites/217331979/plans/`.
Could it be this issue is a race condition? Some checks that prevent the request from happening because it believes the internet is not ready yet.
Do we have a retry strategy if something goes wrong with the fetching plan request?

Here is the video:

https://user-images.githubusercontent.com/18119390/233134319-669ea959-6739-49e2-bc17-d3d7bf79e6de.mp4

**Comment by wzieba** [comment]
This issue is really strange - even though we both have Pixels, our phones behave differently. It's especially interesting that just after "offline banner" is hidden on your phone, you receive "Error fetching data" snackbar. It doesn't happen for me 🤔 

https://user-images.githubusercontent.com/5845095/233604916-0000c0e1-cefb-4569-813f-d656f648a8d8.mp4

**Comment by wzieba** [comment]
> Could it be this issue is a race condition? Some checks that prevent the request from happening because it believes the internet is not ready yet.

I took this lead - might be very valid, as `NetworkInfo` reports connection to a network - not to the internet.

I've added a 1-sec delay, so the system can have time to "realize" if it's connected to the internet really. Could you please test it? Thanks 🙏

**Comment by atorresveiga** [comment]
> I've added a 1-sec delay, so the system can have time to "realize" if it's connected to the internet really. Could you please test it? Thanks 🙏

With the delay, now the code is working great. The problem seemed to be the speed at which our devices connected to the internet.

I'm worried that 1s might not work for every device. I would like to suggest adding a backoff retry strategy when the error is a network error (not being connected to the internet), something that could be tackled in a different PR, so I'll pre-approve this one and let t you decide.

**Review by atorresveiga — APPROVED** [review]
Great work @wzieba! Your use of unit tests boosted my understanding of the code's logic and increased our code base's quality. 👏 👌

LGTM! :shipit:

**Comment by wzieba** [comment]
> I'm worried that 1s might not work for every device. I would like to suggest adding a backoff retry strategy when the error is a network error (not being connected to the internet), something that could be tackled in a different PR, so I'll pre-approve this one and let t you decide.

You're right. AFAIK the only really reliable strategy for checking the internet connection would be a ping to a server. But considering the importance of this feature (low), I agree we can handle it in different PR and different time - e.g. when internet connection will be crucial to fulfill some business requirment.

Thanks for the review again!

**Pull request merged**